### PR TITLE
fix `Cannot type a card number into the "Recurly hosted fields" demo` (close #1445)

### LINKED ISSUE #1445: Cannot type a card number into the "Recurly hosted fields" demo
https://dev.recurly.com/docs/recurly-js-getting-started#section-individual-card-fields
Without proxy:
image
With proxy:
image

### PR DESCRIPTION
**Pull request by LavrovArtem**

### CODE REVIEW

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@LavrovArtem  rebase, please

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
FPR

File: src/client/sandbox/native-methods.js [hunk 1]
```diff
@@ -193,9 +193,13 @@ class NativeMethods {
 
         if (win.MessageEvent) {
```
Comment (miherlosev) [inline src/client/sandbox/native-methods.js]: Which of supported browsers did not have `MessageEvent`. Write comment about it.

File: src/client/sandbox/native-methods.js [hunk 2]
```diff
@@ -193,9 +193,13 @@ class NativeMethods {
...
             const originPropDescriptor = win.Object.getOwnPropertyDescriptor(window.MessageEvent.prototype, 'origin');
+            const dataPropDescriptor   = win.Object.getOwnPropertyDescriptor(window.MessageEvent.prototype, 'data');
 
             if (originPropDescriptor && originPropDescriptor.get && originPropDescriptor.configurable)
```
Comment (miherlosev) [inline src/client/sandbox/native-methods.js]: For which of supported browsers this condition equals `false`. Wrtie comment.

File: src/client/sandbox/code-instrumentation/properties/index.js [hunk 3]
```diff
@@ -208,11 +209,20 @@ export default class PropertyAccessorsInstrumentation extends SandboxBase {
             },
 
             data: {
-                condition: el => domUtils.isDomElement(el) && domProcessor.isUrlAttr(el, 'data'),
+                // NOTE: The data property of the MessageEvent object cannot be redefined in the Android 5.1 browser
+                condition: el => domUtils.isDomElement(el) && domProcessor.isUrlAttr(el, 'data') ||
+                                 isAndroid && !nativeMethods.messageEventDataGetter && domUtils.isMessageEvent(el),
```
Comment (miherlosev) [inline src/client/sandbox/code-instrumentation/properties/index.js]: Move `isAndroid && !nativeMethods.messageEventDataGetter && domUtils.isMessageEvent(el)` to the separate utils function `isSupportedMessageEvent`

File: src/client/sandbox/event/message.js [hunk 4]
```diff
@@ -40,8 +41,23 @@ export default class MessageSandbox extends SandboxBase {
...
 
+    _getMessageData (e) {
+        const rawData = nativeMethods.messageEventDataGetter && isMessageEvent(e)
+            ? nativeMethods.messageEventDataGetter.call(e)
+            : e.data;
+
+        return typeof rawData === 'string' ? parseJSON(rawData) : rawData;
+    }
+
+    _callOriginListener (originListener, e) {
```
Comment (miherlosev) [inline src/client/sandbox/event/message.js]: Move to util function (used in 2 places) and rename to `callEventListener` 
https://github.com/DevExpress/testcafe-hammerhead/blob/e4a4096b8d99dfb3a65333276f10f061589a6edf/src/client/sandbox/event/listeners.js#L74
https://github.com/DevExpress/testcafe-hammerhead/blob/cc241be3cfae15f31a0d9295f41a064f53810e39/src/client/sandbox/event/message.js#L69

File: src/client/sandbox/event/message.js [hunk 5]
```diff
@@ -55,21 +71,27 @@ export default class MessageSandbox extends SandboxBase {
     }
 
     _onWindowMessage (e, originListener) {
-        const data = typeof e.data === 'string' ? parseJSON(e.data) : e.data;
+        const isOverriddenEventObject = e[IS_OVERRIDDEN_EVENT_OBJECT];
```
Comment (miherlosev) [inline src/client/sandbox/event/message.js]: Drop `isOverriddenEventObject` and rename internal property to `HAS_OVERRIDEN_DATA_PROP`

File: src/client/utils/dom.js [hunk 6]
```diff
@@ -499,6 +499,13 @@ export function isWindow (instance) {
         return true;
 
     try {
+        // NOTE: A window object received from MessageEvent.target has strange behavior into js debugger:
```
Comment (miherlosev) [inline src/client/utils/dom.js]: Change comment to:

This is special hack for `MSEdge` and `IE` browsers.
`InstanceToString` call result has a strange values for a `MessageEvent.target` property:
`[object DispHTMLWindow2]` for IE11
`[object Object]` for `MSEdge`.

Comment (miherlosev) [reply src/client/utils/dom.js]: Add separate tests for `isWindow` function

File: src/client/utils/dom.js [hunk 7]
```diff
@@ -632,6 +639,10 @@ export function isWebSocket (ws) {
     return instanceToString(ws) === '[object WebSocket]';
 }
 
+export function isMessageEvent (e) {
```
Comment (miherlosev) [inline src/client/utils/dom.js]: Add separate test for this function

File: test/client/fixtures/sandbox/event/message-test.js [hunk 8]
```diff
@@ -391,3 +393,34 @@ test('should not raise an error for sendServiceMessage if window.top is a cross-
             window.removeEventListener('message', onMessageHandler);
         });
 });
+
+test('the "message" event object should be correctly overridden (GH-1445)', function () {
```
Comment (miherlosev) [inline test/client/fixtures/sandbox/event/message-test.js]: `MessageEvent  should be correctly overridden (GH-1445)`

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
FPR

File: src/client/sandbox/native-methods.js [hunk 9]
```diff
@@ -191,16 +191,15 @@ class NativeMethods {
...
 
-        if (win.MessageEvent) {
-            const originPropDescriptor = win.Object.getOwnPropertyDescriptor(window.MessageEvent.prototype, 'origin');
-            const dataPropDescriptor   = win.Object.getOwnPropertyDescriptor(window.MessageEvent.prototype, 'data');
+        const originPropDescriptor = win.Object.getOwnPropertyDescriptor(window.MessageEvent.prototype, 'origin');
+        const dataPropDescriptor   = win.Object.getOwnPropertyDescriptor(window.MessageEvent.prototype, 'data');
 
-            if (originPropDescriptor && originPropDescriptor.get && originPropDescriptor.configurable)
-                this.messageEventOriginGetter = originPropDescriptor.get;
+        // NOTE: This conditions are used for the Android 5.1 browser
```
Comment (miherlosev) [inline src/client/sandbox/native-methods.js]: `These conditions...`

File: src/client/sandbox/event/message.js [hunk 10]
```diff
@@ -4,16 +4,17 @@ import nativeMethods from '../native-methods';
...
+import { callEventListener } from '../../utils/event';
 import fastApply from '../../utils/fast-apply';
-import createEvent from '../../utils/create-event';
 
 const MESSAGE_TYPE = {
     service: 'hammerhead|service-msg',
     user:    'hammerhead|user-msg'
 };
 
+const HAS_OVERRIDEN_DATA_PROP = 'hammerhead|message-event|overridden';
```
Comment (miherlosev) [inline src/client/sandbox/event/message.js]: const HAS_OVERRIDEN_DATA_PROP = 'hammerhead|message-event|has-overridden-data-prop';

File: test/client/fixtures/sandbox/event/message-test.js [hunk 11]
```diff
@@ -391,3 +393,34 @@ test('should not raise an error for sendServiceMessage if window.top is a cross-
...
+
+            try {
+                JSON.stringify(eventObj);
+
+                ok(true);
+            }
+            catch (e) {
+                // NOTE: Browser Android 5.1 cannot stringify a native "message" event.
+                // It fails with 'Converting circular structure to JSON' error.
+                // In the Android 6.0, browser version is 44
```
Comment (miherlosev) [inline test/client/fixtures/sandbox/event/message-test.js]: // In the Android 6.0, browser version is 44, it works.

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
FPR

**Comment by LavrovArtem** [comment]
ping @churkin

**Pull request merged**